Add pre/post processors to allow nan and inf values to be stored in JSON

zhubonan opened the pull request:
This PR allow nan/inf values to be stored as JSON strings in PostgreSQL.

sphuber committed:
Add the `MsonableData` data plugin

The `MSONable` class, provided by the `monty` library, is designed to
make arbitrary classes easily JSON-serializable such that they can be
serialized and, for example, stored in a database. The class is used
extensively in `pymatgen` and many of their classes are MSONable. Since
the `pymatgen` classes are used often in the current AiiDA community, it
would be nice if these objects can be easily stored in the provenance
graph.

The new `MsonableData` data plugin, wraps any instance of a `MSONable`
class. When constructed, it calls `as_dict` which by definition returns
a JSON-serialized version of the object, which we can therefore store in
the nodes attributes and allow it to be stored. The `obj` property will
return the original MSONable instance as it deserializes it from the
serialized version stored in the attributes.

sphuber commented:
Thanks @zhubonan . I'm out of office this week but will take a look when I'm back

sphuber committed:
Add support for float constants infinity and nan

zhubonan committed:
Simplification for the MsonableData class

Remove the subclass checking - only check for the as_dict and from_dict
methods. Remove unecessary round trips for json encoding. We still
assume that the output of `as_dict` is JSON serializable as originally
intended by the `MSONable` class.

sphuber commented:
I have updated the PR to make a more generic plugin, i.e. to decouple it from `MSONable` and allow any class with the necessary behavior. Therefore it also made more sense to me to rename the class and call it `JsonableData`.
